### PR TITLE
Make printing of big(NaN) consistent #23382

### PR DESCRIPTION
**Pull request by jw3126**
Make printing of big(NaN) consistent #23382

### CODE REVIEW

File: test/numbers.jl [hunk 1]
```diff
@@ -3024,9 +3024,14 @@ end
     end
 end
 
-@testset "compact NaN printing" begin
-    @test sprint(io->show(IOContext(io, :compact => true), NaN16)) == "NaN"
-    @test sprint(io->show(IOContext(io, :compact => true), NaN32)) == "NaN"
-    @test sprint(io->show(IOContext(io, :compact => true), NaN64)) == "NaN"
-    @test_broken sprint(io->show(IOContext(io, :compact => true), big(NaN))) == "NaN"
+@testset "printing non finite floats" for T in subtypes(AbstractFloat)
```
Comment (StefanKarpinski) [inline test/numbers.jl]: I like the use of `subtypes(AbstractFloat)` here. Slick.

File: test/numbers.jl [hunk 2]
```diff
@@ -3024,9 +3024,14 @@ end
...
+    for (x, sx) in [(T(NaN), "NaN"),
+                    (-T(NaN), "NaN"),
+                    (T(Inf), "Inf"),
+                    (-T(Inf), "-Inf")]
+        @assert x isa T
+        @test string(x) == sx
+        @test sprint(io -> show(IOContext(io, :compact => true), x)) == sx
+        @test sprint(print, x) == sx
+    end
```
Comment (StefanKarpinski) [inline test/numbers.jl]: Not really relevant to this PR, but I wonder if we check the non-compact printing of these anywhere...

**Review by StefanKarpinski — APPROVED** [review]

File: test/numbers.jl (hunk 2 above)
Comment (jw3126) [inline test/numbers.jl]: I don't think so. `grep "\"NaN" . -r` gives:
```
./numbers.jl:@test repr(NaN) == "NaN"
./numbers.jl:@test repr(-NaN) == "NaN"
./numbers.jl:@test repr(NaN32) == "NaN32"
./numbers.jl:@test repr(-NaN32) == "NaN32"
./numbers.jl:    for (x, sx) in [(T(NaN), "NaN"),
./numbers.jl:                    (-T(NaN), "NaN"),
./printf.jl:@test (@sprintf "%f" NaN) == "NaN"
./printf.jl:@test (@sprintf "%f" big"NaN") == "NaN"
./float16.jl:@test repr(NaN16) == "NaN16"
./float16.jl:@test sprint(showcompact, NaN16) == "NaN"
./resolvedata1.jl:    ["DualNumbers", v"0.1.3", "NaNMath"],
./resolvedata1.jl:    ["DualNumbers", v"0.1.5", "NaNMath"],
./resolvedata1.jl:    ["DualNumbers", v"0.2.2", "NaNMath"],
./resolvedata1.jl:    ["DualNumbers", v"0.3.0", "NaNMath"],
./resolvedata1.jl:    ["ForwardDiff", v"0.2.5", "NaNMath"],
./resolvedata1.jl:    ["ForwardDiff", v"0.3.4", "NaNMath", v"0.2.2"],
./resolvedata1.jl:    ["ForwardDiff", v"0.3.5", "NaNMath", v"0.2.1"],
./resolvedata1.jl:    ["ForwardDiff", v"0.4.2", "NaNMath", v"0.2.2"],
./resolvedata1.jl:    ["Graphics", v"0.2.0", "NaNMath", v"0.2.4"],
./resolvedata1.jl:    ["NaNMath", v"0.1.1", "Compat"],
./resolvedata1.jl:    ["NaNMath", v"0.2.0"],
./resolvedata1.jl:    ["NaNMath", v"0.2.1"],
./resolvedata1.jl:    ["NaNMath", v"0.2.3", "Compat", v"0.9.1"],
./resolvedata1.jl:    ["NaNMath", v"0.2.4", "Compat", v"0.9.1"],
./resolvedata1.jl:    "NaNMath"=>v"0.2.4",
```

Comment (StefanKarpinski) [inline test/numbers.jl]: Ok, seems like we could be a little more complete about testing our NaN printing.

**Comment by StefanKarpinski** [comment]
Why do you keep closing and opening this?

**Comment by jw3126** [comment]
There were random CI failures. I wanted it to rerun until everything passes. Is there a better way to do this?

**Comment by fredrikekre** [comment]
That restarts succesful runs too, which is a waste of CI resources. Better to ask someone to restart the specific runs that failed.

**Comment by iblis17** [comment]
seem bsd ci keep freezing on this PR?
I will rebuild 2 or more times again.

**Comment by iblis17** [comment]
https://julia.iblis.cnmc.tw/#/builders/1/builds/1658
the `bitarray` testsuit is missing from ci report

---

*will keep update this post*.
(*maybe I should make a bot report missing testsuit automatically.*)

**Comment by jw3126** [comment]
```
Error in testset statistics:

Error During Test

  Got an exception of type LoadError outside of a @test
  LoadError: return type RowVector{Float64,Array{Float64,1}} does not match inferred return type Any
  Stacktrace:
   [1] setindex!(::Array{StackFrame,1}, ::StackFrame) at ./abstractarray.jl:0
  while loading /tmp/julia/share/julia/test/statistics.jl, in expression starting on line 177
ERROR: LoadError: Test run finished with errors
while loading /tmp/julia/share/julia/test/runtests.jl, in expression starting on line 29
```
Comes up quite often. I don't see how that is connected to my changes to `BigFloat` printing. Any ideas?

**Comment by fredrikekre** [comment]
> Comes up quite often.

That's #23371

**Comment by jw3126** [comment]
Can someone rerun circleci and review this PR?

**Comment by jw3126** [comment]
I think this PR is ready.

File: base/mpfr.jl [hunk 3]
```diff
@@ -894,30 +894,63 @@ end
...
+    ccall((:mpfr_snprintf,:libmpfr),
+        Int32, (Ptr{UInt8}, Culong, Ptr{UInt8}, Ptr{BigFloat}...),
+        buf, 0, fmt, &x)
+end
+
+function _fill_buffer!(buf, fmt, x::BigFloat)
+    s = length(buf)
+    # we temporarily need one more item in buffer to capture null termination
+    resize!(buf, s + 1)
+    n = ccall((:mpfr_sprintf,:libmpfr), Int32, (Ptr{UInt8}, Ptr{UInt8}, Ptr{BigFloat}...), buf, fmt, &x)
```
Comment (musm) [inline base/mpfr.jl]: This is using thr deprecated cccall syntax

**Pull request merged**